reactionrank command

cchollar opened the pull request:
Never done this before but here goes!

I have added my /reactionrank command to the bot and done some testing to ensure it handles failures correctly. Please review the below code and let me know if it passes inspection! Thank you =).

cchollar committed:
Added reactionrank command to bot w/ refactor, added constants to cfg.py

cchollar committed:
Emoji names now text, added comments

cchollar committed:
Added nojoy option

gsfernandes81 commented on ionic/__main__.py:
```diff
@@ -1,15 +1,18 @@
 import lightbulb as lb
 import regex as re
 import sqlalchemy as sql
+from sqlalchemy.sql.traversals import ColIdentityComparatorStrategy
```
This doesn't seem to be used @cchollar ,
Is there any reason this is here?

cchollar replied:
I have no idea why this is here, my pyright install must have tried to put it in lol I will get this removed

gsfernandes81 replied:
No worries!

gs729 commented:
I've made the mistake of leaving in 3 of my commits in dev without pushing them, so will merge those into main first, merge in your commit ec40037 since that isn't related to the reaction rank command and then look at merging this once we're both happy with the code

gsfernandes81 commented on ionic/__main__.py:
```diff
@@ -1,15 +1,19 @@
 import argparse
 import asyncio
 import datetime as dt
+from datetime import datetime, timedelta
```
Could we switch to using dt.datetime and dt.timedelta everywhere since the code already imports datetime as dt just above it?

cchollar replied:
Absolutely, python namespaces had me confused. I will remove the line and swap to dt.datetime and dt.timedelta

gsfernandes81 replied:
Much appreciated!

gsfernandes81 commented on ionic/__main__.py:
```diff
@@ -788,6 +495,124 @@ async def on_lb_start(event: lb.LightbulbStartedEvent):
 
 
+@bot.command()
+@lb.option("nojoy", "Ignores :rofl: and :joy: reactions",bool, default=True)
+@lb.option(
+    "hours",
+    "How far back should I pull messages from (int)",
+    int,
+    default=6,
+    max_value=24,
+    min_value=1,
+)
+@lb.option("channel", "Target channel to search", h.TextableGuildChannel, default=None)
```
Would maybe switching around the order of these be better for the UX?

Having channel first seems like it might be better, happy to hear your opinions on this of course

cchollar replied:
I have a similar opinion to yours, so I will swap the order of the options!

gsfernandes81 commented on ionic/__main__.py:
```diff
@@ -788,6 +495,124 @@ async def on_lb_start(event: lb.LightbulbStartedEvent):
+@lb.option("channel", "Target channel to search", h.TextableGuildChannel, default=None)
+@lb.command("reactionrank", "List top 5 reactions in specified channel")
+@lb.implements(lb.SlashCommand)
+async def ReactionRank(ctx: lb.Context) -> None:
```
[PEP 0008](https://peps.python.org/pep-0008/#function-and-variable-names) conventions typically have functions named "lowercase, with words separated by underscores".

Would you be happy to change this accordingly? @cchollar

cchollar replied:
Sure thing, I'm not entirely sure why I didn't use PEP 0008 conventions in the first place,

cchollar committed:
Changes from code review for reaction rank PR

gsfernandes81 committed:
Black reformat cfg.py

cchollar committed:
Reaction rank now deletes keys with 0 emoji count, pizza servers only

cchollar committed:
Merge branch 'dev' of https://github.com/gs729/ionic-traces into dev

cchollar committed:
Black format __main__.py

cchollar commented:
I had to do a git pull to get the latest version of dev after [e2cfa1d]. I'm not sure if I did it right by using a git merge. Let me know if I should have done something different!! Nevertheless, I have changed the command to only work on the pizza_servers list. And have also added a small control flow that checks the reaction dictionary for any keys with a count subkey of 0, and removes them. This should stop the bot from displaying its own emoji in edge cases.

gs729 commented:
Thanks for the fix!

I'll be moving these changes to a feature branch "feature-reactionrank", squashing the commits and closing this pull request. 

I'll let you open a PR with that branch to main so that it's attributed to you. I've made a few changes to the code including removing the use of Decimal(...) and getcontext(...) and instead using round(...) and moving long inline comments to their own lines but other than that this is still all your code and I've marked you as co-author in the squashed commit

We can move our conversation to the new PR once you make it from feature-reactionrank to dev.
